src/sagemaker_xgboost_container/algorithm_mode/callback.py

```diff
@@ -0,0 +1,205 @@
+FILE_LOCK_SUFFIX = ".sagemaker-uploading"
+FILE_SAFE_SUFFIX = ".sagemaker-uploaded"
+
+CHECKPOINT_FILENAME = "xgboost-checkpoint"
+CHECKPOINT_NUM_DIGITS = 12
+
+logger = logging.getLogger(__name__)
+logging.basicConfig(level=logging.INFO)
+
+
+def _zero_pad(iteration: int) -> str:
+
+    iter_num = format(iteration, "0{}".format(CHECKPOINT_NUM_DIGITS))
+
+    return iter_num
+
+
+# modified from https://github.com/dmlc/xgboost/blob/master/python-package/xgboost/callback.py
+def print_evaluation(period=1, show_stdv=True, start_iteration=0):
+    """Create a callback that print evaluation result.
+
+    We print the evaluation results every **period** iterations
+    and on the first and the last iterations.
+
+    Parameters
+    ----------
+    period : int
+        The period to log the evaluation results
+
+    show_stdv : bool, optional
+         Whether show stdv if provided
+
+    Returns
+    -------
+    callback : function
+        A callback that print evaluation every period iterations.
+    """
+    def callback(env):
+        """internal function"""
+        if env.rank != 0 or (not env.evaluation_result_list) or period is False or period == 0:
+            return
+        i = env.iteration
+        if i % period == 0 or i + 1 == env.begin_iteration or i + 1 == env.end_iteration:
+            msg = '\t'.join([fmt_metric(x, show_stdv) for x in env.evaluation_result_list])
+            rabit.tracker_print('[%d]\t%s\n' % (i + start_iteration, msg))
+    return callback
+
+
+class SaveCheckpoint:
+    """Create a callback that saves checkpoints to disk.
+
+    This class represents the hook which is meant to be used a callback
+    function in XGBoost.
+
+    Example
+    -------
+    >>> save_checkpoint = SaveCheckpoint("/opt/ml/checkpoints")
+    >>> xgboost.train(prams, dtrain, callbacks=[save_checkpoint])
+
+    At the end of each iteration, we save each checkpoint to different files
+    parametrized by the iteration number, e.g., checkpoint.01, checkpoint.02,
+    etc. However, this could result in a large number of files to save locally
+    (and on S3 if used in SageMaker). To save disk space and reduce the amount
+    of data required to be downloaded on resumption, we retain only the N
+    (default 5, spcified by 'max_to_keep') most recent checkpoints.
+
+    To delete stale checkpoints, we use a producer-consumer pattern: we start a
+    daemon thread in the background and maintain a queue of files to delete.
+    There may be a lock on the file if SageMaker is uploading the file; in that
+    case, the file is put back on the queue and we try again later. When
+    training is complete, we put SENTINEL on the queue, and when we see the
+    SENTINEL, we clean up and exit the thread.
+    """
+
+    SENTINEL = None
+
+    def __init__(
+            self, checkpoint_dir: str, max_to_keep: int = 5, start_iteration: int = 0,
+            num_round: Optional[int] = None
+            ) -> None:
```

Comment:
Ending parens should be in the same line as last element, see pep: https://www.python.org/dev/peps/pep-0008/#indentation